src/collapse/dtrace.rs

```diff
@@ -113,19 +120,129 @@ impl Collapse for Folder {
-        Self {
-            cache_inlines: Vec::new(),
-            occurrences: FnvHashMap::with_capacity_and_hasher(512, fnv::FnvBuildHasher::default()),
-            stack: VecDeque::default(),
-            stack_str_size: 0,
-            opt,
+impl Folder {
+    fn collapse_single_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        let mut line = String::with_capacity(CAPACITY_LINE_BUFFER);
+
+        // skip header lines -- first empty line marks start of data
+        loop {
+            line.clear();
+            if reader.read_line(&mut line)? == 0 {
+                // We reached the end :( this should not happen.
+                warn!("File ended while skipping headers");
+                return Ok(());
+            };
+            if line.trim().is_empty() {
+                break;
+            }
+        }
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            let line = line.trim();
+
+            if line.is_empty() {
+                continue;
+            } else if let Ok(count) = line.parse::<usize>() {
+                self.on_stack_end(count);
+            } else {
+                self.on_stack_line(line);
+            }
+        }
+        Ok(())
+    }
+
+    fn collapse_multi_threaded<R>(&mut self, mut reader: R) -> io::Result<()>
+    where
+        R: io::BufRead,
+    {
+        debug_assert!(self.occurrences.is_concurrent());
+        debug_assert!(self.opt.nthreads > 1);
+
+        let mut buf = Vec::with_capacity(CAPACITY_INPUT_BUFFER);
+        reader.read_to_end(&mut buf)?;
+
+        let mut input = Input::new(buf, self.opt.nthreads, Self::identify_stack_locations)?;
+
+        crossbeam::thread::scope(|scope| {
+            let mut handles = Vec::with_capacity(input.nthreads());
+            let (sender, receiver) = crossbeam::channel::bounded(input.nthreads());
+            for chunk in input.chunks() {
+                let occurrences = self.occurrences.clone();
+                let opt = self.opt;
+                let sender = sender.clone();
+
+                let handle = scope.spawn(move |_| {
+                    let mut folder = Folder {
+                        // state
+                        cache_inlines: Vec::new(),
+                        occurrences,
+                        stack: VecDeque::default(),
+                        stack_str_size: 0,
+
+                        // options
+                        opt,
+                    };
+                    let result = folder.collapse_single_threaded(chunk);
+                    sender.send(result).unwrap();
+                });
+                handles.push(handle);
+            }
+            for handle in handles {
+                receiver.recv().unwrap()?;
+                handle.join().unwrap();
+            }
+            Ok::<_, io::Error>(())
+        })
+        .unwrap()?;
```

Comment:
You could probably save yourself some type annotation here by removing the `Ok(())` line below and the semicolon here.